sc/shared/src/main/scala/sigmastate/lang/SigmaTyper.scala

```diff
@@ -48,475 +48,488 @@ class SigmaTyper(val builder: SigmaBuilder,
                  bound: SValue,
-                 expected: Option[SType] = None): SValue = ( bound match {
-    case Block(bs, res) =>
-      var curEnv = env
-      val bs1 = ArrayBuffer[Val]()
-      for (v @ Val(n, _, b) <- bs) {
-        if (curEnv.contains(n)) error(s"Variable $n already defined ($n = ${curEnv(n)}", v.sourceContext)
-        val b1 = assignType(curEnv, b)
-        curEnv = curEnv + (n -> b1.tpe)
-        builder.currentSrcCtx.withValue(v.sourceContext) {
-          bs1 += mkVal(n, b1.tpe, b1)
-        }
-      }
-      val res1 = assignType(curEnv, res)
-      mkBlock(bs1.toSeq, res1)
-
-    case Tuple(items) =>
-      mkTuple(items.map(assignType(env, _)))
-
-    case c @ ConcreteCollection(items, _) =>
-      val newItems = items.map(assignType(env, _))
-      assignConcreteCollection(c, newItems)
-
-    case i @ Ident(n, _) =>
-      env.get(n) match {
-        case Some(t) => mkIdent(n, t)
-        case None =>
-          SGlobal.method(n) match {
-            case Some(method) if method.stype.tDom.length == 1 => // this is like  `groupGenerator` without parentheses
-              val srcCtx = i.sourceContext
-              processGlobalMethod(srcCtx, method, IndexedSeq())
-            case _ =>
-              error(s"Cannot assign type for variable '$n' because it is not found in env $env", bound.sourceContext)
-          }
-      }
-
-    case sel @ Select(obj, n, None) =>
-      val newObj = assignType(env, obj)
-      newObj.tpe match {
-        case tNewObj: SProduct =>
-          val iField = tNewObj.methodIndex(n)
-          val method = if (iField != -1) {
-            tNewObj.methods(iField)
-          } else
-            throw new MethodNotFound(s"Cannot find method '$n' in in the object $obj of Product type with methods ${tNewObj.methods}", obj.sourceContext.toOption)
-          val tMeth = method.stype
-          val tRes = tMeth match {
-            case SFunc(args, _, _) =>
-              val tThis = args(0)  // first arg corresponds to method's receiver
-              val tMethSpec = unifyTypes(tThis, tNewObj) match {
-                case Some(subst) if subst.nonEmpty => applySubst(tMeth, subst).asFunc // specialize for concrete receiver type
-                case _ => tMeth.asFunc
-              }
-              if (tMethSpec.tDom.length == 1 && tMethSpec.tpeParams.isEmpty) tMethSpec.tRange
-              else tMethSpec.copy(tDom = tMethSpec.tDom.tail, tRange = tMethSpec.tRange)
-            case _ => tMeth
+                 expected: Option[SType] = None): SValue = {
+    val typedValue = ( bound match {
+      case Block(bs, res) =>
+        var curEnv = env
+        val bs1 = ArrayBuffer[Val]()
+        for (v @ Val(n, givenTpe, b) <- bs) {
+          if (curEnv.contains(n)) error(s"Variable $n already defined ($n = ${curEnv(n)}", v.sourceContext)
+          val expectedTpe = if (givenTpe != NoType) Some(givenTpe) else None
+          val b1 = assignType(curEnv, b, expectedTpe)
```

Comment:
Passing `givenTpe` from the `Val` node and passing it to `assignType` recursively.